orderbook/src/database/app_data.rs

```diff
@@ -0,0 +1,225 @@
+use super::*;
+use anyhow::Result;
+use futures::stream::TryStreamExt;
+use futures::StreamExt;
+use model::app_data::AppDataBlob;
+use primitive_types::{H160, H256};
+use std::borrow::Cow;
+use thiserror::Error;
+
+#[async_trait::async_trait]
+pub trait AppDataStoring: Send + Sync {
+    async fn insert_app_data(&self, app_data: &AppDataBlob) -> Result<H256, InsertionError>;
+    fn app_data<'a>(&'a self, filter: &'a AppDataFilter) -> BoxStream<'a, Result<AppDataBlob>>;
+}
+
+/// Any default value means that this field is unfiltered.
+#[derive(Clone, Default)]
+pub struct AppDataFilter {
+    pub app_data_hash: Option<H256>,
+    pub referrer: Option<H160>,
+}
+
+#[derive(Error, Debug)]
+pub enum InsertionError {
+    #[error("Parsing the string was not successful:")]
+    ParsingStringError(#[from] serde_json::Error),
+    #[error("Duplicated record for `{0}`:")]
+    DuplicatedRecord(H256),
+    #[error("Anyhow error:")]
+    AnyhowError(#[from] anyhow::Error),
+    #[error("Database error:")]
+    DbError(#[from] sqlx::Error),
+}
+
+#[async_trait::async_trait]
+impl AppDataStoring for Postgres {
+    async fn insert_app_data(&self, app_data_str: &AppDataBlob) -> Result<H256, InsertionError> {
+        let app_data = app_data_str.get_app_data()?;
+        const QUERY: &str = "\
+            INSERT INTO app_data (\
+                app_data_hash, app_code, referrer, file_blob)\
+                VALUES ($1, $2, $3, $4);";
+        let app_data_hash = app_data_str.sha_hash()?;
+        sqlx::query(QUERY)
+            .bind(app_data_hash.clone().as_bytes())
+            .bind(
+                app_data
+                    .app_code
+                    .unwrap_or_else(|| "".to_string())
+                    .as_bytes(),
+            )
+            .bind(
+                app_data
+                    .metadata
+                    .unwrap_or_default()
+                    .referrer
+                    .unwrap_or_default()
+                    .address
+                    .as_ref(),
```

Comment:
It looks like here you will insert the 0x0000 address by default. Is that intended? Shouldn't we just insert None if any of the optional fields are not set?

Reply:
I agree, if the column is nullable, we should probably insert null (i.e. `None`). Nice catch!